google/cloud/documentai_toolbox/wrappers/document_revision.py

```diff
@@ -0,0 +1,346 @@
+                        Document2 = {pages:List[Page],entities:List[Entity],revision_nodes:List[RevisionNodes]...}
+    """
+
+    revised_entities, history = _modify_docproto(revision.entities)
+    return revised_entities, history
+
+
+def get_level(doc: DocumentWithRevisions):
+    try:
+        level = 0
+        p = doc.parent
+        if p.revision_id:
+            while p:
+                p = p.parent
+                level += 1
+        return level
+    except:
+        return 0
+
+
+def _print_child_tree(
+    currect_revision: DocumentWithRevisions, doc: DocumentWithRevisions, seen
+):
+    if currect_revision.revision_id == doc.revision_id:
+        print(" " * get_level(doc) + "*|--", end="")
+    else:
+        print("  " * get_level(doc) + "|--", end="")
+    print(doc.revision_id)
+    if doc.children:
+        for each in doc.children:
+            seen.append(each.revision_id)
+            _print_child_tree(currect_revision, each, seen)
+
+
+@dataclasses.dataclass
+class DocumentWithRevisions:
+    r"""Represents a wrapped Document.
+
+    A single Document protobuf message might be written as several JSON files on
+    GCS by Document AI's BatchProcessDocuments method.  This class hides away the
+    shards from the users and implements convenient methods for searching and
+    extracting information within the Document.
+
+    Attributes:
+        gcs_prefix (str):
+            Required.The gcs path to a single processed document.
+
+            Format: `gs://{bucket}/{optional_folder}/{operation_id}/{folder_id}`
+                    where `{operation_id}` is the operation-id given from BatchProcessDocument
+                    and `{folder_id}` is the number corresponding to the target document.
+    """
+
+    document: Document = dataclasses.field(init=True, repr=False)
+    revision_nodes: List[documentai.Document] = dataclasses.field(init=True, repr=False)
+    gcs_prefix: str = dataclasses.field(init=True, repr=False, default=None)
+    parent_ids: List[str] = dataclasses.field(init=True, repr=False, default=None)
+
+    next_: Document = dataclasses.field(init=False, repr=False, default=None)
+    last: Document = dataclasses.field(init=False, repr=False, default=None)
+    revision_id: str = dataclasses.field(init=False, repr=False, default=None)
+    history: List[str] = dataclasses.field(init=False, repr=False, default_factory=list)
+    root_revision: Document = dataclasses.field(init=False, repr=False, default=None)
+
+    parent: DocumentWithRevisions = dataclasses.field(
+        init=False, repr=False, default=None
+    )
+
+    children: List[DocumentWithRevisions] = dataclasses.field(
+        init=False, repr=False, default_factory=list
+    )
+    children_ids: List[str] = dataclasses.field(
+        init=False, repr=False, default_factory=list
+    )
+    root_revision_nodes: List[DocumentWithRevisions] = dataclasses.field(
+        init=False, repr=False, default_factory=list
+    )
```

Comment:
Ok, I guess I understand more why the extra class for revisions is here (to keep the base Document class simpler) but I'm still not sure I like this solution, it seems like there's a lot of fields involved and this object could get quite large with all of the Documents and `DocumentWithRevisions` included.

Reply:
This is implemented like this for multiple reasons, the first one is to keep continuity with the internal implementation. Another the document that we are importing from gcs here is completely different we are not taking in json files we're processing `dp.bp`  files which is binary protobuf and the revision documents should have a tree like hierarchy which would introduce a lot of clutter/confusion in Document. So it's best to separate it out to make it easier to make changes to `DocumentWithRevision`. Also there are a lot of fields to support the tree structure and to support movement between non-children nodes.